### ISSUE TITLE
API Question: can we make this even more versatile?

### ISSUE THREAD

**Issue by silkentrance** [neurosnap/gen-readlines#9]
Considering the following and making this more streamlined to existing node fs/stream APIs...

E.g.

`file` path as first parameter followed by a second parameter `options`

```
readlines(file, options)
```

where options would then include

```
{
    offset: optional offset into the input file (aka position)
    fd: optional file descriptor
    encoding: the default being UTF-8
}
```

Of course, in order to support a different encoding, the overall logic needs some additional tuning.

The result would, of course, be that readlines would now depend upon the various node APIs for decoding input data.

Furthermore, the user of the API will no longer have to fs.open... and then fs.stat... as that would all be inherent to readlines.

This could be taken further by accepting a decode handler that will decode input data, e.g. encrypted data and so on.

What do you think?

**Comment by neurosnap** [comment]
Originally I went this route but I didn't want to presume how the end-dev would open the file.  Having said that if we can preserve the option for them to supply fd/size information while making it easier to interact with this library then I'm all for it.

**Comment by silkentrance** [comment]
I do not know, but does node support file descriptors for arbitrary resources, e.g. http/s, ftp/s, and so on.? If so, then your approach is definitely the better solution.

And even if so, most of them would be mapped to an EventEmitter based approach and I am still struggling with making this a generator based interface.

My initial approaches of mapping a event emitter to a generator all failed and I cannot think of any indirection that would be able to make this a generator based interface unless implementing something native. Yet, again, I am getting older :grin:

So basically, the user must be able to provide an additional option key, namely `filesize`?

**Comment by silkentrance** [comment]
@neurosnap 
For starters, I have looked into deasync and wrapping line-by-line so that readlines() would yield the lines it gets from the underlying LineByLineReader. Turns out that this will add way too much overhead and all optimizations made will be mute. 

I reckon that since gen-readlines makes no assumption on the encoding at all except for \r and \n being single bytes, this issue is moot and the user can either take the buffer as is or use iconv or other such APIs to decode the data in a second step.

Closing.
